forum/communications/tests.py

```diff
@@ -1,3 +1,199 @@
+        try:
+            self._enter_chat_room(room_1)
+            self._open_new_window()
+            self._enter_chat_room(room_2)
+
+            # Post message in first window
+            self._switch_to_window(0)
+            message_1 = "hello"
+            self._post_message(message_1)
+            self._wait_for_message(message_1)
+
+            # Post message in second window
+            self._switch_to_window(1)
+            message_2 = "world"
+            self._post_message(message_2)
+            self._wait_for_message(message_2)
+
+            # Verify message visibility
+            if expected_shared:
+                self.assertIn(
+                    message_1,
+                    self._chat_log_value,
+                    f"Message '{message_1}' not shared correctly.",
+                )
+            else:
+                self.assertNotIn(
+                    message_1,
+                    self._chat_log_value,
+                    f"Message '{message_1}' improperly shared between rooms.",
+                )
+        finally:
+            self._close_all_new_windows()
+
+    def _enter_chat_room(self, room_name):
+        self.driver.get(f"{self.live_server_url}/api/v1/communications/")
+        room_input = WebDriverWait(self.driver, 10).until(
+            EC.presence_of_element_located((By.ID, "room-name-input")),
+            "Room input field not found",
+        )
+        room_input.clear()
+        room_input.send_keys(room_name, Keys.ENTER)
+        WebDriverWait(self.driver, 10).until(
+            lambda driver: room_name in driver.current_url,
+            "Failed to navigate to the chat room",
+        )
+
+    def _open_new_window(self):
+        self.driver.execute_script('window.open("about:blank", "_blank");')
+        self._switch_to_window(-1)
+
+    def _close_all_new_windows(self):
+        while len(self.driver.window_handles) > 1:
+            self._switch_to_window(-1)
+            self.driver.close()
+        self._switch_to_window(0)
+
+    def _switch_to_window(self, window_index):
+        self.driver.switch_to.window(self.driver.window_handles[window_index])
+
+    def _post_message(self, message):
+        chat_input = WebDriverWait(self.driver, 10).until(
+            EC.presence_of_element_located((By.ID, "chat-message-input")),
+            "Chat message input field not found",
+        )
+        chat_input.clear()
+        chat_input.send_keys(message, Keys.ENTER)
+
+    def _wait_for_message(self, message):
+        WebDriverWait(self.driver, 10).until(
+            lambda driver: message in self._chat_log_value,
+            f"Message '{message}' was not received",
+        )
+
+    @property
+    def _chat_log_value(self):
+        chat_log = WebDriverWait(self.driver, 10).until(
+            EC.presence_of_element_located((By.ID, "chat-log")),
+            "Chat log element not found",
+        )
+        return chat_log.get_property("value")
```

Comment:
- Add tests for WebSocket disconnection and reconnection scenarios.
 - Test sending invalid JSON payloads over WebSocket.
 - Validate the handling of edge cases, such as attempting to join a non-existent room.
 
_Examples:_
```
def test_websocket_reconnection(self):
    """Test that WebSocket can reconnect after an unexpected closure."""
    self._enter_chat_room("room_1")
    self._switch_to_window(0)
    self.driver.execute_script("window.location.reload()")
    WebDriverWait(self.driver, 10).until(
        lambda driver: "room_1" in driver.current_url,
        "Failed to reconnect to the WebSocket room after reload",
    )

def test_send_invalid_json(self):
    """Test that sending invalid JSON data raises an error."""
    self._enter_chat_room("room_1")
    self.driver.execute_script("""
        const socket = new WebSocket('ws://' + window.location.host + '/ws/api/v1/communication/room_1/');
        socket.onopen = () => socket.send('Invalid JSON');
    """)
    error_message = WebDriverWait(self.driver, 10).until(
        lambda driver: "Unexpected error" in driver.page_source,
        "Error message not displayed for invalid JSON",
    )
    self.assertTrue(error_message)
```

Reply:
Done